### ISSUE TITLE
Running GUI applications within Kata Containers accelerated with iGLX

### ISSUE THREAD

**Issue by mviereck** [kata-containers/kata-containers#3524]
Follow up from https://github.com/kata-containers/runtime/issues/586

It is possible now to run hardware accelerated GUI applications in kata containers using indirect rendering (iGLX) with Xorg.
iGLX has been broken a long time in Xorg, but is fixed since Xorg version 1.20.8.

An example with [x11docker](https://github.com/mviereck/x11docker) v7.0.1 and `kata-runtime  : 1.13.0-alpha0`:
```
x11docker --runtime=kata-runtime --xorg --gpu --network x11docker/check glxspheres64
``` 

Drawbacks:
 - Indirect rendering is less performant than direct rendering. If sharing GPU device files and X unix sockets from host would be possible, the performance could be severely improved.
 - A few applications seem to have issues with iGLX. At least `glxgears` does not work properly.

Other supported options of x11docker for kata containers:
 - `--pulseaudio` for sound.
 - `--printer` to access CUPS printers.

**Comment by Jakob-Naucke** [comment]
I should think this depends on #722.

**Comment by mviereck** [comment]
Oh, ok. I wasn't aware that the 2.x series does not provide an OCI runtime anymore.

**Comment by mviereck** [comment]
Thank you!
The tickets are interesting.

For x11docker the most straightforward way is to use `nerdctl`. It already supports it with option `--backend=nerdctl`. `ctr` and `crictl` have a quite restricted cli, they are not practical for x11docker setups.

However,  `nerdctl` (current version 0.16.0) fails with `--runtime=containerd-shim-kata-v2` (cgroupv2 is enabled):
```
$ sudo nerdctl run --rm --runtime=containerd-shim-kata-v2 busybox hostname
FATA[0000] failed to create shim: OCI runtime create failed: unable to retrieve OCI runtime error (open /run/containerd/io.containerd.runtime.v2.task/default/b26f993d7fb8fdcc82769e136476d6e647609e1ce7a01daf6fe8b14cced769a2/log.json: no such file or directory): containerd-shim-kata-v2 did not terminate successfully: exit status 2: unknown 

$ sudo nerdctl run --rm -ti --runtime=containerd-shim-kata-v2 busybox hostname
FATA[0000] failed to create shim: OCI runtime create failed: unable to retrieve OCI runtime error (open /run/containerd/io.containerd.runtime.v2.task/default/57208652012267c18b9906b7991f42ff2780731cdee81d1e955948232ae4c1cc/log.json: no such file or directory): containerd-shim-kata-v2 did not terminate successfully: exit status 2: flag provided but not defined: -root
Usage of containerd-shim-kata-v2:
  -address string
    	grpc address back to main containerd
  -bundle string
    	path to the bundle if not workdir
  -debug
    	enable debug output in logs
  -id string
    	id of the task
  -namespace string
    	namespace that owns the shim
  -publish-binary string
    	path to publish binary (used for publishing events) (default "containerd")
  -socket string
    	socket path to serve
  -v	show the shim version and exit
: unknown 
```
I've installed kata 2.3.1 with `snapd`:
```
$ containerd-shim-kata-v2 --version
Kata Containers containerd shim: id: "io.containerd.kata.v2", version: 2.3.1, commit: 365e358115824beb055a64767dc23c46dfcd8813
```
I am not sure if the bug is on side of kata or nerdctl. The first try looks like a kata bug, the second one like a nerdctl bug.

Once `nerdctl --runtime=containerd-shim-kata-v2 [...]` works, it should be possible to run accelerated GUIs in kata as described above. One would just need to add `--backend=nerdctl --runtime=containerd-shim-kata-v2` to the x11docker command. 
For now I might just wait for further development until nerdctl+kata works well.

**Comment by Jakob-Naucke** [comment]
@mviereck I shall take a look, Kata+nerdctl was definitely working at one point. We're not running that in our CI or anything though.

**Comment by Jakob-Naucke** [comment]
@mviereck You need reverse domain notation, i.e. `io.containerd.kata.v2` rather than `containerd-shim-kata-v2`.

**Comment by Jakob-Naucke** [comment]
@mviereck yeah, our attention on standalone support hasn't been so good, sorry :) both volumes and network will work under k8s, so these things aren't entirely broken with Kata, but I understand that this isn't ideal for everyone. BTW, probably the latest issue on network with standalone tools is #2580. FYI you will need a working CNI even if we didn't have this issue.

**Comment by mviereck** [comment]
> yeah, our attention on standalone support hasn't been so good, sorry :) 

No problem :-)
I'd just like to support kata beside the other already working "classic" runtimes. There is no urgent need on my side.
I was once asked to support `kata-runtime` (that still works), and I would like to continue the support with kata 2.x.
I have subscribed to the related tickets and will give the setup another try once they are solved.

> both volumes and network will work under k8s

I am not sure yet what k8s means, some special or future version? The glossary does not mention it.
https://github.com/kata-containers/kata-containers/wiki/Glossary#k

Thanks again for looking at this!

**Comment by Jakob-Naucke** [comment]
> I am not sure yet what k8s means, some special or future version? The glossary does not mention it. https://github.com/kata-containers/kata-containers/wiki/Glossary#k

Oops, got lost in jargon there :) k8s is short for [Kubernetes](https://kubernetes.io/).

We can keep this open in my opinion because I _almost_ fear iGLX won't work out of the box. Because this depends on other issues, I'm going to label this as `needs-more-info`. Let me know if you disagree.

**Comment by mviereck** [comment]
> Oops, got lost in jargon there :) k8s is short for Kubernetes.

Thank you for clarification!

> We can keep this open in my opinion because I almost fear iGLX won't work out of the box. Because this depends on other issues, I'm going to label this as needs-more-info. Let me know if you disagree.

We can keep this open, of course. I am happy if this is still of interest.

I don't expect issues with iGLX because it does not need hardware access within the container. All it needs is a network connection to the host. The rendering is done by the X server on host. (For clarification, iGLX means indirect rendering done by an X server accessed over TCP, in opposite to usual direct rendering that needs hardware access by the X client in container.)

The file sharing issue is not related to this, that is just an issue for x11docker that always shares some files with the container.

**Comment by JayDoubleu** [comment]
So with katav1 no longer available on fedora i was wondering if it would make sense to bring it back as if this doesn't get implemented we're effectively without any kata containers support for podman. 

Would it be reasonable to say that the v1 kata-containers package would work with podman if built manually?
